File: cpp/src/DO/Sara/Graphics/ImageDraw.cpp
// ========================================================================== //
// This file is part of Sara, a basic set of libraries in C++ for computer
// vision.
//
// Copyright (C) 2013-2016 David Ok <david.ok8@gmail.com>
//
// This Source Code Form is subject to the terms of the Mozilla Public
// License v. 2.0. If a copy of the MPL was not distributed with this file,
// you can obtain one at http://mozilla.org/MPL/2.0/.
// ========================================================================== //

#include <QtGui>

#include <DO/Sara/Graphics.hpp>
#include <DO/Sara/Graphics/GraphicsUtilities.hpp>


namespace DO { namespace Sara {

  auto draw_point(ImageView<Rgb8>& image, int x, int y, const Rgb8& c) -> void
  {
    QImage qimage(as_QImage(image));
    QPainter p(&qimage);
    p.setPen(to_QColor(c));
    p.drawPoint(x, y);
  }

  auto draw_line(ImageView<Rgb8>& image, int x1, int y1, int x2, int y2,
                 const Rgb8& c, int pen_width, bool antialiasing) -> void
  {
    QImage qimage(as_QImage(image));
    QPainter p(&qimage);
    p.setRenderHints(QPainter::Antialiasing, antialiasing);
    p.setPen(QPen(to_QColor(c), pen_width));
    p.drawLine(x1, y1, x2, y2);
  }

  auto draw_rect(ImageView<Rgb8>& image, int x, int y, int w, int h,
                 const Rgb8& c, int pen_width) -> void
  {
    QImage qimage(as_QImage(image));
    QPainter p(&qimage);
    p.setPen(QPen(to_QColor(c), pen_width));
    p.drawRect(x, y, w, h);
  }

  auto draw_circle(ImageView<Rgb8>& image, int xc, int yc, int r, const Rgb8& c,
                   int pen_width, bool antialiasing) -> void
  {
    QImage qimage(as_QImage(image));
    QPainter p(&qimage);
    p.setRenderHints(QPainter::Antialiasing, antialiasing);
    p.setPen(QPen(to_QColor(c), pen_width));
    p.drawEllipse(QPoint(xc, yc), r, r);
  }

  auto draw_ellipse(ImageView<Rgb8>& image, const Eigen::Vector2f& center,
                    float r1, float r2, float degree, const Rgb8& c,
                    int pen_width, bool antialiasing) -> void
  {
    auto qimage = as_QImage(image);

    QPainter p(&qimage);
    p.setRenderHints(QPainter::Antialiasing, antialiasing);
    p.setPen(QPen(to_QColor(c), pen_width));
    p.translate(QPointF(center.x(), center.y()));
    p.rotate(degree);
    p.translate(-r1, -r2);
    p.drawEllipse(QRectF(0, 0, 2 * r1, 2 * r2));
  }

  auto draw_arrow(ImageView<Rgb8>& image, int x1, int y1, int x2, int y2,
                  const Rgb8& col, int pen_width, int arrow_width,
                  int arrow_height, int style, bool antialiasing) -> void
  {
    auto qimage = as_QImage(image);

    QPainter p(&qimage);
    p.setRenderHints(QPainter::Antialiasing, antialiasing);

    const auto qcol = to_QColor(col);
    auto sl = qreal{};
    const qreal dx = x2 - x1;
    const qreal dy = y2 - y1;
    const qreal norm = qSqrt(dx * dx + dy * dy);
    if (norm < 0.999)  // null vector
    {
      p.setPen(QPen(qcol, pen_width));
      p.drawPoint(x1, y1);
      return;
    }

    auto path = QPainterPath{};
    auto pts = QPolygonF{};

    const auto dx_norm = dx / norm;
    const auto dy_norm = dy / norm;
    const auto p1x =
        x1 + dx_norm * (norm - arrow_width) + arrow_height / 2. * dy_norm;
    const auto p1y =
        y1 + dy_norm * (norm - arrow_width) - arrow_height / 2. * dx_norm;
    const auto p2x =
        x1 + dx_norm * (norm - arrow_width) - arrow_height / 2. * dy_norm;
    const auto p2y =
        y1 + dy_norm * (norm - arrow_width) + arrow_height / 2. * dx_norm;
    switch (style)
    {
    case 0:
      p.setPen(QPen(qcol, pen_width));
      p.drawLine(x1, y1, x2, y2);
      p.drawLine(x2, y2, int(p1x), int(p1y));
      p.drawLine(x2, y2, int(p2x), int(p2y));
      break;
    case 1:
      pts << QPointF(p2x, p2y);
      pts << QPointF(x2, y2);
      pts << QPointF(p1x, p1y);
      sl = norm - (arrow_width * .7);
      pts << QPointF(x1 + dx_norm * sl + dy_norm * pen_width,
                     y1 + dy_norm * sl - dx_norm * pen_width);
      pts << QPointF(x1 + dy_norm * pen_width, y1 - dx_norm * pen_width);
      pts << QPointF(x1 - dy_norm * pen_width, y1 + dx_norm * pen_width);
      pts << QPointF(x1 + dx_norm * sl - dy_norm * pen_width,
                     y1 + dy_norm * sl + dx_norm * pen_width);
      path.addPolygon(pts);
      p.fillPath(path, qcol);
      break;
    case 2:
      pts << QPointF(p2x, p2y);
      pts << QPointF(x2, y2);
      pts << QPointF(p1x, p1y);
      sl = norm - arrow_width;
      pts << QPointF(x1 + dx_norm * sl + dy_norm * pen_width,
                     y1 + dy_norm * sl - dx_norm * pen_width);
      pts << QPointF(x1 + dy_norm * pen_width, y1 - dx_norm * pen_width);
      pts << QPointF(x1 - dy_norm * pen_width, y1 + dx_norm * pen_width);
      pts << QPointF(x1 + dx_norm * sl - dy_norm * pen_width,
                     y1 + dy_norm * sl + dx_norm * pen_width);
      path.addPolygon(pts);
      p.fillPath(path, qcol);
      break;
    default:
      break;
    }
  }

  auto draw_text(ImageView<Rgb8>& image, int x, int y, const std::string& text,
                 const Rgb8& color, int font_size, float orientation,
                 bool italic, bool bold, bool underline, int pen_width,
                 bool antialiasing) -> void
  {
    auto qimage = as_QImage(image);

    QPainter p(&qimage);
    p.setRenderHints(QPainter::Antialiasing, antialiasing);

    auto font = QFont{};
    font.setPointSize(font_size);
    font.setItalic(italic);
    font.setBold(bold);
    font.setUnderline(underline);

    auto textPath = QPainterPath{};
    const auto baseline = QPointF{0, 0};
    textPath.addText(baseline, font,
                     QString::QString::fromLocal8Bit(text.c_str()));

    // Outline the text by default for more visibility.
    p.setBrush(to_QColor(color));
    p.setPen(QPen(Qt::black, pen_width));
    p.setFont(font);

    p.translate(x, y);
    p.rotate(static_cast<qreal>(orientation));
    p.drawPath(textPath);
  }

  auto fill_rect(ImageView<Rgb8>& image, int x, int y, int w, int h,
                 const Rgb8& c) -> void
  {
    QImage qimage(as_QImage(image));
    QPainter p(&qimage);
    p.fillRect(x, y, w, h, to_QColor(c));
  }

  auto fill_circle(ImageView<Rgb8>& image, int x, int y, int r, const Rgb8& c)
      -> void
  {
    QImage qimage(as_QImage(image));
    QPainter p(&qimage);
    QPainterPath path;
    path.addEllipse(QPointF(x, y), qreal(r), qreal(r));
    p.fillPath(path, to_QColor(c));
  }

}}  // namespace DO::Sara

File: python/oddkiva/sara/graphics/image_draw.py
# Copyright (C) 2025 David Ok <david.ok8@gmail.com>

from typing import Iterable

import numpy as np

from PySide6.QtCore import QPointF, QRectF, QSizeF
from PySide6.QtGui import (
    QBrush,
    QColor,
    QFont,
    QFontMetrics,
    QImage,
    QPainter,
    QPen,
)


def to_qimage(array: np.ndarray):
    h, w, c = array.shape
    qimage = QImage(array.data, w, h, c * w, QImage.Format.Format_RGB888)
    return qimage

def draw_point(array: np.ndarray, x, y, color, antialiasing: bool = True):
    """ Draw a point.
    """
    surface = to_qimage(array)
    p = QPainter(surface)
    p.setRenderHints(QPainter.RenderHint.Antialiasing, antialiasing)
    p.setPen(QColor(*color));
    p.drawPoint(x, y);
    p.end()

def draw_line(array: np.ndarray, p1, p2, color, pen_width,
              antialiasing: bool = True):
    """ Draw a line.
    """
    surface = to_qimage(array)
    p = QPainter(surface)
    p.setRenderHints(QPainter.RenderHint.Antialiasing, antialiasing)
    p.setPen(QPen(QColor(*color), pen_width))
    p.drawLine(QPointF(*p1), QPointF(*p2))
    p.end()

def draw_rect(array: np.ndarray, top_left_corner, sizes, color, pen_width,
              antialiasing: bool = True):
    """ Draw a rectangle.
    """
    surface = to_qimage(array)
    p = QPainter(surface)
    p.setRenderHints(QPainter.RenderHint.Antialiasing, antialiasing)
    p.setPen(QPen(QColor(*color), pen_width))
    p.drawRect(top_left_corner[0], top_left_corner[1],
                           sizes[0], sizes[1]);
    p.end()

def draw_circle(array: np.ndarray, center, radius, color, pen_width,
                antialiasing: bool = True):
    """ Draw a circle.
    """
    surface = to_qimage(array)
    p = QPainter(surface)
    p.setRenderHints(QPainter.RenderHint.Antialiasing, antialiasing)
    p.setPen(QPen(QColor(*color), pen_width))
    p.drawEllipse(QPointF(*center), radius, radius);
    p.end()

def draw_ellipse(array: np.ndarray, center, r1, r2, angle_in_degrees: float,
                 color: Iterable[int], pen_width: int,
                 antialiasing: bool =True):
    """ Draw an ellipse.
    """
    surface = to_qimage(array)
    p = QPainter(surface)
    p.setRenderHints(QPainter.RenderHint.Antialiasing, antialiasing)
    p.save()
    p.setPen(QPen(QColor(*color), pen_width))
    p.translate(QPointF(*center))
    p.rotate(angle_in_degrees)
    p.translate(-r1, -r2)
    p.drawEllipse(QRectF(0, 0, 2 * r1, 2 * r2))
    p.restore()
    p.end()

def draw_text(array: np.ndarray, p: Iterable[int], text: str,
              color: Iterable[int], font_size: int, orientation: float,
              italic: bool, bold: bool, underline: bool,
              antialiasing: bool = True):
    """ Draw a string.
    """
    font = QFont()
    font.setPointSize(font_size)
    font.setItalic(italic)
    font.setBold(bold)
    font.setUnderline(underline)

    surface = to_qimage(array)
    painter = QPainter(surface)
    painter.setRenderHints(QPainter.RenderHint.Antialiasing, antialiasing)

    painter.save()
    painter.setPen(QColor(*color))
    painter.setFont(font)

    painter.translate(p[0], p[1])
    painter.rotate(orientation)
    painter.drawText(0, 0, text)
    painter.restore()
    painter.end()

def draw_boxed_text(array: np.ndarray,
                    p: Iterable[float],
                    text: str,
                    box_color: Iterable[int | float],
                    font: QFont,
                    text_color: Iterable[int | float] = (0, 0, 0),
                    angle: float = 0.) -> None:
    font_metrics = QFontMetrics(font)
    text_offset = 1
    def calculate_text_box_size(text: str):
        label_text_rect = font_metrics.boundingRect(text)
        size = label_text_rect.size()
        w, h = size.width() + text_offset * 2 + 1, size.height()
        return w, h

    w, h = calculate_text_box_size(text)
    xoff = -5
    yoff = h * 0.5
    l, t = (int(p[0] + xoff + 0.5), int(p[1] + 0.5 + yoff))
    fill_rect(array, (l - text_offset, int(t - h + 0.5)), (w, h), box_color)
    draw_text(array, (l, t - 2 * text_offset), text, text_color,
              font.pointSize(), angle,
              font.italic(), font.bold(), font.underline())

def draw_image(array: np.ndarray, image: np.ndarray,
               offset: Iterable[int | float],
               scale: float):
    """ Draw an image.
    """
    surface = to_qimage(array)
    p = QPainter(surface)

    xoff, yoff = offset
    p.translate(xoff, yoff)
    p.scale(scale, scale)
    p.drawImage(0, 0, to_qimage(image))
    p.scale(1 / scale, 1 / scale)
    p.translate(-xoff, -yoff)
    p.end()


def fill_rect(array: np.ndarray,
              top_left_corner: Iterable[int],
              sizes: Iterable[int],
              color: Iterable[int | float]):
    surface = to_qimage(array)
    p = QPainter(surface)

    p.begin(surface)
    p.fillRect(
        QRectF(QPointF(*top_left_corner), QSizeF(*sizes)),
        QBrush(QColor(*color))
    )
    p.end()
